pkg/controller/clusterserviceversion/clusterserviceversion_controller.go

```diff
@@ -131,16 +165,47 @@ func (r *ReconcileClusterServiceVersion) Reconcile(request reconcile.Request) (r
 		return reconcile.Result{}, err
 	}
 
+	annotations := CSV.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// check if CSV is being deleted
+	// if yes -> finalizer logic
+	// if no -> do nothing
+	if !CSV.ObjectMeta.DeletionTimestamp.IsZero() && utils.Contains(CSV.GetFinalizers(), utils.CSV_FINALIZER) {
+		//Run finalization logic for the CSV.
+		result, isReconcile, err := r.finalizeCSV(CSV)
+		if isReconcile {
+			return result, err
+		}
```

Comment:
```suggestion
		}
		return reconcile.Result{}, nil
```
If finalizer has run with no reconciles, then we should return a finished result.